### PR TITLE
Handle joining to an existing cluster

### PR DESCRIPTION
**Pull request by jsravn**

### CODE REVIEW

**Comment by jsravn** [comment]
Let me merge, please.

File: lib/bootstrap.go [hunk 1]
```diff
@@ -43,19 +43,45 @@ func (b *bootstrapper) Bootstrap() string {
...
+
+	instanceURLs := make(map[string]bool)
+	for _, instance := range b.asg.GetInstances() {
+		instanceURLs[peerURL(instance.PrivateIP)] = true
+	}
+
+	for memberURL := range memberURLs {
+		if !instanceURLs[memberURL] {
+			log.Infof("%s not found in ASG, removing member from etcd", memberURL)
+			b.cluster.RemoveMember(memberURL)
```
Comment (totahuanocotl) [inline lib/bootstrap.go]: what will happen if the member can't be removed? 
Will the whole bootstrapping fail? or is the error ignored?

Comment (jsravn) [inline lib/bootstrap.go]: You mean if it's already removed? I believe it gets a 409 from etcd and succeeds.

**Comment by jsravn** [comment]
Please re-review, I've changed it a lot as the original version wasn't working when I tested it.

File: lib/bootstrap.go [hunk 2]
```diff
@@ -43,17 +43,51 @@ func (b *bootstrapper) Bootstrap() string {
...
-	members := b.cluster.Members()
-	return len(members) == 0
+	m := b.cluster.MemberURLs()
+	return len(m) == 0
 }
 
 func (b *bootstrapper) bootstrapNew() string {
-	return b.createEtcdConfig(newCluster)
+	var availURLs []string
+	for _, i := range b.asg.GetInstances() {
```
Comment (sebbonnet) [inline lib/bootstrap.go]: This piece of code can be reused. Retrieving the available urls is done when reconcilingMembers

File: lib/etcdcluster/etcdcluster.go [hunk 3]
```diff
@@ -35,18 +39,55 @@ func New(asg asg.ASG) Cluster {
...
+	members := ensureGetMembers(membersAPI)
+
+	for _, member := range members {
+		assertSinglePeerURL(member)
+		if member.PeerURLs[0] == peerURL {
+			return membersAPI.Remove(context.Background(), member.ID)
+		}
+	}
+
+	return nil
```
Comment (sebbonnet) [inline lib/etcdcluster/etcdcluster.go]: May be useful to log a message when the member appears already removed.

Comment (jsravn) [inline lib/etcdcluster/etcdcluster.go]: Ok, I had that before but removed it.

**Pull request merged**